### PR TITLE
apiserver: further cleanup of apiserver storage plumbing

### PR DESCRIPTION
**Pull request by sttts**
- move kubeapiserver`s `RESTOptionsFactory` back to EtcdOptions by adding a `AddWithStorageFactoryTo`
- factor out storage backend `Config` construction from EtcdOptions
- move all `StorageFactory` related code into server/storage subpackage.

In short: remove my stomach ache about `kubeapiserver.RESTOptionsFactory`.

approved based on #40363

### CODE REVIEW

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -68,7 +69,7 @@ type ServerRunOptions struct {
 func NewServerRunOptions() *ServerRunOptions {
 	s := ServerRunOptions{
 		GenericServerRunOptions: genericoptions.NewServerRunOptions(),
-		Etcd:                 genericoptions.NewEtcdOptions(kubeoptions.DefaultEtcdPathPrefix, api.Scheme, nil),
+		Etcd:                 genericoptions.NewEtcdOptions(storagebackend.NewDefaultConfig(kubeoptions.DefaultEtcdPathPrefix, api.Scheme, nil)),
```
Comment (deads2k) [inline cmd/kube-apiserver/app/options/options.go]: I'd really like keep the arguments to `NewRecommendedOptions` to be: "Where do you want it and what's your scheme".  Even if we construct this under the covers.  I haven't gotten to the point to see if you plumbed it through that way or not, I just thought of it here.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 2]
```diff
@@ -114,29 +108,46 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
 }
 
 func (s *EtcdOptions) ApplyTo(c *server.Config) error {
-	c.RESTOptionsGetter = &restOptionsFactory{options: s}
-
+	c.RESTOptionsGetter = RESTOptionsFactoryFunc(func(resource schema.GroupResource) (generic.RESTOptions, error) {
+		ret := generic.RESTOptions{
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: how does this `ResourcePrefix` get set?

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: If we missed this and just broke it, update the sample API server test to prove storage and actually check its location in  etcd.  We've had pain related to messing this up in openshift.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 3]
```diff
@@ -114,29 +108,46 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
...
-		ResourcePrefix:          f.options.StorageConfig.Prefix + "/" + resource.Group + "/" + resource.Resource,
-	}
-
-	if !f.options.EnableWatchCache {
-		ret.Decorator = generic.UndecoratedStorage
-	}
+type RESTOptionsFactoryFunc func(resource schema.GroupResource) (generic.RESTOptions, error)
 
-	return ret, nil
+func (f RESTOptionsFactoryFunc) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: I really dislike this construct.  Can we meet halfway?  The rest of the pull makes it, but there's a struct implementing the interface and you build that?

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_codec.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/recognizer"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+
+	"github.com/golang/glog"
+)
+
+// StorageCodecConfig are the arguments passed to newStorageCodecFn
+type StorageCodecConfig struct {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_codec.go]: straight move?

**Comment by deads2k** [comment]
I'm largely ok with it.  A few requests on expression.

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_codec.go (hunk 4 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_codec.go]: yes

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: It's idiomatic, compare net/http handlers. But, I am fine with the struct if you hate it so much.

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: > It's idiomatic, compare net/http handlers. But, I am fine with the struct if you hate it so much.

I really do.  Many of the "idiomatic" things we've done in the code base have resulted in messes.  I don't feel bad about breaking from idiomatic approach.

File: staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go [hunk 5]
```diff
@@ -23,20 +23,40 @@ import (
...
+	v1GroupVersion = schema.GroupVersion{Group: "", Version: "v1"}
+
 	registry = registered.NewOrDie(os.Getenv("KUBE_API_VERSIONS"))
 	announce = make(announced.APIGroupFactoryRegistry)
+	scheme         = runtime.NewScheme()
+	codecs         = serializer.NewCodecFactory(scheme)
+	parameterCodec = runtime.NewParameterCodec(scheme)
 )
 
 func init() {
```
Comment (MHBauer) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go]: why two init blocks?

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory_test.go]: No good reason, will fix.

File: cmd/kube-apiserver/app/options/options.go (hunk 1 above)
Comment (sttts) [inline cmd/kube-apiserver/app/options/options.go]: I don't like that. We construct the storagebackend config behind the scenes and use it from `EtcdOptions` to create `RESTOptionsFactories`. How do you want to customize the backend? Patching the config after creating the EtcdOptions is dirty, especially if you pass prefix, scheme and codec before. Using NewDefaultConfig is using clean composition. What's the issue with that?

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 2 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: Fixed.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: fixed. Not pretty.

**Comment by sttts** [comment]
Adressed all comments. @deads2k will squash if you confirm that you really really really want those two revert commits. Both are ugly IMO.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 6]
```diff
@@ -115,47 +115,54 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
...
-
-		return ret, nil
-	})
+	c.RESTOptionsGetter = &storageFactoryRestOptionsFactory{EtcdOptions: *s, StorageFactory: factory}
 	return nil
 }
 
-type RESTOptionsFactoryFunc func(resource schema.GroupResource) (generic.RESTOptions, error)
+type simpleRestOptionsFactory struct {
+	EtcdOptions
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: don't like anonymous includes.

**Comment by deads2k** [comment]
Yeah, I prefer `Revert functional style one-func interface`.  You can remove `Revert storagebackend.NewDefaultConfig` as long as the storageconfig doesn't leak into `RecommendedOptions`

**Comment by sttts** [comment]
@k8s-bot cvm gke e2e test this

**Pull request merged**